metals/src/main/scala/scala/meta/internal/decorations/SyntheticDecorationProvider.scala

```diff
@@ -0,0 +1,240 @@
+      Future.successful(())
+  }
+
+  def onChange(textDocument: TextDocuments, path: Path): Unit = {
+    for {
+      focused <- focusedDocument()
+      filePath = AbsolutePath(path)
+      sourcePath <- SemanticdbClasspath.toScala(workspace, filePath)
+      if sourcePath == focused || !clientConfig.isDidFocusProvider()
+      textDoc <- textDocument.documents.headOption
+      source <- fingerprints.loadLastValid(sourcePath, textDoc.md5, charset)
+    } {
+      val docWithText = textDoc.withText(source)
+      Document.set(docWithText)
+      publishImplicitInfo(sourcePath, docWithText)
+    }
+  }
+
+  private def toSymbolName(symbol: String, textDoc: TextDocument): String = {
+    if (symbol.isLocal)
+      textDoc.symbols
+        .find(_.symbol == symbol)
+        .map(_.displayName)
+        .getOrElse(symbol)
+    else symbol
+  }
+
+  private def toHoverString(textDoc: TextDocument)(symbol: String): String = {
+    toSymbolName(symbol, textDoc)
+      .replace("/", ".")
+      .stripSuffix(".")
+      .stripSuffix("#")
+      .stripPrefix("_empty_.")
+      .replace("#", ".")
+      .replace("()", "")
+  }
+
+  private def toDecorationString(
+      textDoc: TextDocument
+  )(symbol: String): String = {
+    if (symbol.isLocal)
+      textDoc.symbols
+        .find(_.symbol == symbol)
+        .map(_.displayName)
+        .getOrElse(symbol)
+    else symbol.desc.name.value
+  }
+
+  private def decorate(
+      lspRange: l.Range,
+      decorationText: String,
+      hoverMessage: String
+  ) = {
+
+    def decoration(range: l.Range, text: String, hover: l.MarkupContent) =
+      new DecorationOptions(
+        range,
+        hover,
+        ThemableDecorationInstanceRenderOptions(
+          after = ThemableDecorationAttachmentRenderOptions(
+            text,
+            color = "grey",
+            fontStyle = "italic",
+            opacity = 0.7
+          )
+        )
+      )
+
+    val fullMessage = s"""|```scala
+                          |$hoverMessage
+                          |```""".stripMargin
+    // There is an issue hover for VS Code https://github.com/microsoft/vscode/issues/105302
+    val markup = new l.MarkupContent(
+      l.MarkupKind.MARKDOWN,
+      fullMessage
+    )
+
+    // sythetics get reported over multiple lines
+    val start = lspRange.getStart()
+    val end = lspRange.getEnd()
+    if (start.getLine() != end.getLine()) {
+      start.setLine(end.getLine())
+      start.setCharacter(0)
+    }
+    List(decoration(lspRange, decorationText, markup))
+  }
+
+  private def publishDecorations(
+      path: AbsolutePath,
+      textDocument: TextDocument
+  ): Unit = {
+    val decorations = for {
+      synthetic <- textDocument.synthetics
+      range <- synthetic.range.toIterable
+      edit = buffer.tokenEditDistance(path, textDocument.text)
+      decoration <- printSyntheticInfo(
+        textDocument,
+        synthetic,
+        toDecorationString(textDocument),
+        userConfig(),
```

Comment:
I was curious why we passed in `userConfig()` here. TIL that we have `implicitArgumentAnnotationsEnable` and `typeAnnotationsEnabled` settings.

Reply:
This is just for invoking this method, so it doesn't really matter. It will be invoked each time with new user settings.

Reply:
ha, yes, was just a comment when I looked at when the settings were added and say it was a year ago, so I thought they were in there the entire time without me every seeing them. 😆 but as I mentioned before, that was just the odd timestamp, so now worries!